source/drivers/MicroBitQuadratureDecoder.cpp

```diff
@@ -0,0 +1,233 @@
+  *
+  * @param  The maximum interval between samples in microseconds.
+  *
+  * @return MICROBIT_OK on success, or MICROBIT_INVALID_PARAMETER if the configuration is invalid.
+  */
+int MicroBitQuadratureDecoder::setSamplePeriodUs(uint32_t period)
+{
+    if (period < 128)
+        return MICROBIT_INVALID_PARAMETER;
+    samplePeriod = period;
+    return MICROBIT_OK;
+}
+
+/**
+  * Returns the current sampling period.
+  *
+  * @return The sampling period in microseconds.
+  */
+uint32_t MicroBitQuadratureDecoder::getSamplePeriod()
+{
+    return samplePeriod;
+}
+
+/**
+  * Configure the hardware to keep this instance up to date.
+  *
+  * Several instances can exist so long as no more than one of them is
+  * attached to the hardware.  This can be a practical way to control
+  * several motors with their own encoders if they only run at different
+  * times.
+  *
+  * @return MICROBIT_OK on success, MICROBIT_BUSY if the hardware is already attached to another instance, or MICROBIT_INVALID_PARAMETER if the configuration is invalid.
+  */
+int MicroBitQuadratureDecoder::start()
+{
+    int sampleper;
+
+    faults = 0;
+
+    for (sampleper = 7; sampleper >= 0; --sampleper)
+    {
+        // Find the highest (most power-efficient) sample period available
+        // which is not greater than the configuration.  A longer period could
+        // miss input transitions.
+        if ((128u << sampleper) <= samplePeriod)
+            break;
+    }
+
+    if ((status & MICROBIT_COMPONENT_RUNNING) != 0)
+        return MICROBIT_BUSY;
+
+    NRF_QDEC->SHORTS = 0;           // No shorts
+    NRF_QDEC->INTENCLR = ~0;        // No interrupts
+    NRF_QDEC->LEDPOL = (flags & QDEC_LED_ACTIVE_LOW) != 0 ? 0 : 1;
+    NRF_QDEC->SAMPLEPER = sampleper;
+    NRF_QDEC->REPORTPER = 7;        // Slowest possible reporting (not used)
+    NRF_QDEC->PSELLED = LED.name;
+    NRF_QDEC->PSELA = phaseA.name;
+    NRF_QDEC->PSELB = phaseB.name;
+    NRF_QDEC->DBFEN = (flags & QDEC_USING_DEBOUNCE) != 0 ? 1 : 0;
+    NRF_QDEC->LEDPRE = LEDDelay;
+
+    // If these pins were previously triggering events (eg., when emulating
+    // quadrature decoder using transition events) then put a stop to that.
+    if (LED.name != NC)
+        LED.eventOn(MICROBIT_PIN_EVENT_NONE);
+    phaseA.eventOn(MICROBIT_PIN_EVENT_NONE);
+    phaseB.eventOn(MICROBIT_PIN_EVENT_NONE);
+
+    // This is what all the cool kids are doing, so I'll do it too.
+    __NOP();
+    __NOP();
+    __NOP();
+
+    NRF_QDEC->TASKS_READCLRACC = 1; // Clear accumulators
+    NRF_QDEC->ENABLE = 1;
+
+    NRF_QDEC->TASKS_START = 1;
+    status |= MICROBIT_COMPONENT_RUNNING;
+
+    if ((flags & QDEC_USING_SYSTEM_TICK) != 0)
+        system_timer_remove_component(this);
+
+    return MICROBIT_OK;
+}
+
+/**
+  * Stop the hardware and make it available for use by other instances.
+  */
+void MicroBitQuadratureDecoder::stop()
+{
+    if ((flags & QDEC_USING_SYSTEM_TICK) != 0)
```

Comment:
I think I've named the flag badly.  It should be `QDEC_USE_SYSTEM_TICK`.  It's a configuration item that comes in through the constructor or through `disable`/`enableSystemTick()`.  `stop()` should just obey the flag and unregister the hook if it's currently registered.

Perhaps it's more appropriate to keep a flag in `status` confirming when the hook is registered, rather than registering and unregistering it based on what the configuration implies that we should have done in the past.  Though I'm not sure if the extra complexity offers any real value.